wellcomecollection/wellcomecollection.org
# Stop special opening times box displaying against all venues when of type `other`

## What does this change?

Relates to #10640 
I'm hoping I understood the mechanisms well enough, please do challenge if not.

- I split `VenueHours` in multiple components as I would come to reuse `JauntyBox` (great name), as well as helping me read through the component. I know this makes the changes harder to see so happy to run through the PR together.
- Modified the Cardigan story so the Exceptional hours could be seen on there. 
   Because the "shared exceptional hours" (e.g. bank holidays) depend on fetching Prismic data and that doesn't work on Cardigan, I mocked some data in the component itself. 
- ["opening times" helpers](https://github.com/wellcomecollection/wellcomecollection.org/pull/11097/files#diff-e02258b859976bb055b7c1bf2346879c67ab968bdf661a21b366df0c87a8e726): Changed `getOverrideDatesForAllVenues` to filter _out_ exceptional hours of type "other" ("Exceptional") as they are to be displayed only next to their specific venue. I created `getOverrideDatesForSpecificVenue` for those. Both helpers use the newly created `sortOverrideDatesForVenues`, which is just what was done originally in `getOverrideDatesForAllVenues`  but is now needed for both.
- I changed the Layout grid in `[CollectionVenue](https://github.com/wellcomecollection/wellcomecollection.org/pull/11097/files#diff-0868b44701004853e12bb58c0b3673a7d349a3f1c6ad8f6bfc9b3c5e4106f51b)`. This wasn't part of the requirements but it seems like the grid was decided on 5 years ago and I'm thinking used to work a bit better than it does now where there is an indent on the left? 
   <img width="1187" alt="Screenshot 2024-08-09 at 15 34 39" src="https://github.com/user-attachments/assets/239b300e-1d08-423f-bcd5-053cd8f0da17">

  I changed it so it would be more centered
  <img width="1015" alt="Screenshot 2024-08-09 at 15 15 03" src="https://github.com/user-attachments/assets/545d7ff8-4b6d-4267-bf84-0d0012a0ff2c">

  The component is used only in a few pages and the change seemed to not affect where it's used inside a body (e.g. http://localhost:3000/pages/Wuw19yIAAK1Z3Smm) so I think it's alright. Lmk if not!
- I've not touched `content/webapp/components/VenueHours/VenueHours.UnusualOpeningHours.tsx` apart from splitting it in its own file.
- `content/webapp/components/VenueHours/index.tsx`: Now has a separate box for Individual/"other"/exceptional closing hours and those dates are separated from the others through `upcomingExceptionalOpeningHours`
- [Tweaked a test](https://github.com/wellcomecollection/wellcomecollection.org/pull/11097/files#diff-ff23631e58e20a0ac28f028ddfcffd348c2654a041322d058ea55d1df5f05973R130) to check that only non-`other` typed special hours were returned in "common" closures functions 

> [!NOTE]
> I seemed to have needed to upgrade Playwright and its docker container 😄 so that's an extra change that's slipped in there...

## How to test

Run locally
Page with just one venue: http://localhost:3000/pages/Wuw19yIAAK1Z3Smm
Page with all venues: http://localhost:3000/pages/WwQHTSAAANBfDYXU

In Prismic staging, I've added 02/09 to Shop as an special closure of type `other`.
How to run Staging API: 
- Go to `common/services/prismic/fetch/index.ts` and change line 31 to always point to staging.
  `const endpoint = prismic.getRepositoryEndpoint(`wellcomecollection-stage`);`

The messaging should appear on Shop only, as per the ticket rules. 

## How can we measure success?

The opening times page becomes more intuitive

## Have we considered potential risks?

It gets more confusing because something wasn't caught.
